lib/operations/aggregate.js

```diff
@@ -0,0 +1,133 @@
+'use strict';
+
+const AggregationCursor = require('../aggregation_cursor');
+const applyWriteConcern = require('../utils').applyWriteConcern;
+const decorateWithCollation = require('../utils').decorateWithCollation;
+const decorateWithReadConcern = require('../utils').decorateWithReadConcern;
+const handleCallback = require('../utils').handleCallback;
+const MongoError = require('mongodb-core').MongoError;
+const resolveReadPreference = require('../utils').resolveReadPreference;
+const toError = require('../utils').toError;
+
+/**
+ * Perform an aggregate operation. See Collection.prototype.aggregate or Db.prototype.aggregate for more information.
+ *
+ * @method
+ * @param {Db} db A Db instance.
+ * @param {Collection|string} coll A collection instance or the string '1', used for db.aggregate.
+ * @param {object} [pipeline=[]] Array containing all the aggregation framework commands for the execution.
+ * @param {object} [options] Optional settings. See Collection.prototype.aggregate or Db.prototype.aggregate for a list of options.
+ * @param {Db~aggregationCallback|Collection~aggregationCallback} callback The command result callback
+ */
+function aggregate(db, coll, pipeline, options, callback) {
+  let ignoreReadConcern = false;
+
+  if (typeof options.out === 'string') {
+    pipeline.push({ $out: options.out });
```

Comment:
This is mutating the passed pipeline. Do we clone this somewhere further up? If not, I think we should replace this with `pipeline = pipeline.concat([{ $out: options.out }])`